HPCC-19494 New IBYTI logic not working properly

richardkchapman opened the pull request:
Code was assuming meaning in subChannels array that was not actually valid.
Make the assumption valid by allocating subChannels in a more cyclic fashion.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
Debugged through and observed values and code routes being more what I was expecting.

Note that I don't know if this is the ONLY issue with the new IBYTI code, nor does it explain the memory usage issue which suggests that a Roxie with a high rate of ineffective IBYTI is leaking - this needs to be investigated elsewhere.

<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

richardkchapman committed:
HPCC-19494 New IBYTI logic not working properly

Code was assuming meaning in subChannels array that was not actually valid.
Make the assumption valid by allocating subChannels in a more cyclic fashion.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 3aaa983c7e58d1b4a881682fd088403794af7b10
Build: success
Install hpccsystems-platform-community_6.4.16-rc1.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |97 |97 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |734 |734 |0 |
|test (thor) |629 |629 |0 |
|test (roxie) |762 |762 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   13 sec (00:00:13) |  162 sec (00:02:42) |   37 sec (00:00:37) |    6 sec (00:00:06) |   34 sec (00:00:34) | 1052 sec (00:17:32) |   32 sec (00:00:32) | 1336 sec (00:22:16) |

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 1656b3599fdc6e41b7d8264850a88b92e034a81e
Number of Error(s): 1
-- Configuring incomplete, errors occurred!

[Build error:](https://gist.github.com/0c142cb25161bae0f7400d4a37ac6348)
Build log [Build log](https://gist.github.com/HPCCSmoketest/0c142cb25161bae0f7400d4a37ac6348#file-result-18-04-18-16-52-03-log)

richardkchapman committed:
HPCC-19494 New IBYTI logic not working properly

Reduce IBYTI delays more gradually - treating a single timed-out IBYTI as
indicative of a dead node seems to cause far more failed IBYTIs than prior
logic did.

This version should be cleaner and thus easier to understand and fix in the
future, should cope with more than 2 subchannels, and should behave
identically to the pre-6.4.12 code when there are only 2 subchannels.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: a060db770bc7e2a405efe1a0ae76f5c51b0ec7a5
Build: success
Install hpccsystems-platform-community_6.4.16-rc1.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |97 |97 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |734 |734 |0 |
|test (thor) |629 |629 |0 |
|test (roxie) |762 |762 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    6 sec (00:00:06) |  160 sec (00:02:40) |   37 sec (00:00:37) |    6 sec (00:00:06) |   35 sec (00:00:35) | 1055 sec (00:17:35) |   33 sec (00:00:33) | 1332 sec (00:22:12) |

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 92e2189b8f7978b2d8fd6f4a2980f6d1d2c95dc3
Build: success
Install hpccsystems-platform-community_6.4.16-rc1.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |97 |97 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |734 |734 |0 |
|test (thor) |629 |629 |0 |
|test (roxie) |762 |762 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    6 sec (00:00:06) |  166 sec (00:02:46) |   38 sec (00:00:38) |    5 sec (00:00:05) |   33 sec (00:00:33) | 1055 sec (00:17:35) |   34 sec (00:00:34) | 1337 sec (00:22:17) |

ghalliday commented on roxie/ccd/ccdqueue.cpp:
```diff
@@ -608,51 +607,112 @@ void SlaveContextLogger::flush()
-    return (liveSubChannels[channel] & mask) == mask;
-}
+public:
+    unsigned getIbytiDelay(unsigned subChannel) const  // NOTE - zero-based
```
Clearer if the parameter was called primarySubChannel, same for the function below.

ghalliday commented on roxie/ccd/ccdqueue.cpp:
```diff
@@ -1489,9 +1549,10 @@ class RoxieReceiverBase : implements IRoxieOutputQueueManager, public CInterface
             channels[channelCount++] = channel;
         }
+        for (int i = 0; i < channelCount; i++)
+            subChannelInfo[i].init(i);
```
subChannelInfo[channels[i]].init(channels[i]);

or merge into the previous loop

ghalliday commented on roxie/ccd/ccdqueue.cpp:
```diff
@@ -608,51 +607,112 @@ void SlaveContextLogger::flush()
     {
-        unsigned channelMask = SUBCHANNEL_MASK << (SUBCHANNEL_BITS * primarySubChannel);
-        if ((h.retries & ROXIE_RETRIES_MASK) == channelMask)
+        if (primarySubChannel == theirSubChannel)
```
still trying to understand this change...

richardkchapman replied:
It was supposed to make this code clearer! Net effect will be the same.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: a82238e23881376819fec048daf96fd9df3b8b0d
Build: success
Install hpccsystems-platform-community_6.4.16-rc1.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |97 |97 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |734 |734 |0 |
|test (thor) |629 |629 |0 |
|test (roxie) |762 |762 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    5 sec (00:00:05) |  163 sec (00:02:43) |   37 sec (00:00:37) |    6 sec (00:00:06) |   36 sec (00:00:36) | 1070 sec (00:17:50) |   33 sec (00:00:33) | 1350 sec (00:22:30) |
